internal/assets/token_transfer.go

```diff
@@ -298,14 +299,35 @@ func (s *transferSender) sendInternal(ctx context.Context, method sendMethod) er
 
 	switch s.transfer.Type {
 	case fftypes.TokenTransferTypeMint:
-		return plugin.MintTokens(ctx, op.ID, pool.ProtocolID, &s.transfer.TokenTransfer)
+		err = plugin.MintTokens(ctx, op.ID, pool.ProtocolID, &s.transfer.TokenTransfer)
 	case fftypes.TokenTransferTypeTransfer:
-		return plugin.TransferTokens(ctx, op.ID, pool.ProtocolID, &s.transfer.TokenTransfer)
+		err = plugin.TransferTokens(ctx, op.ID, pool.ProtocolID, &s.transfer.TokenTransfer)
 	case fftypes.TokenTransferTypeBurn:
-		return plugin.BurnTokens(ctx, op.ID, pool.ProtocolID, &s.transfer.TokenTransfer)
+		err = plugin.BurnTokens(ctx, op.ID, pool.ProtocolID, &s.transfer.TokenTransfer)
 	default:
 		panic(fmt.Sprintf("unknown transfer type: %v", s.transfer.Type))
 	}
+
+	// if transaction fails,  mark tx and op as failed in DB
+	if err != nil {
+		s.mgr.database.RunAsGroup(ctx, func(ctx context.Context) (err error) {
+			l := log.L(ctx)
+			tx.Status = fftypes.OpStatusFailed
+			if err = s.mgr.database.UpsertTransaction(ctx, tx, false); err != nil {
```

Comment:
You could probably use `UpdateTransaction` instead... although we do almost always use Upsert elsewhere.